clash-cores/src/Clash/Cores/SigmaDeltaADC.hs

```diff
@@ -0,0 +1,262 @@
+  -- 1. the new count of the accumulator
+  -- 2. the downsampled output of the filter and adc
+filterAccumulator accum (count, dataIn, accumulate)
+  | count == 1 && accumulate = (extend dataIn, accum)
+  | accumulate = (accum + extend dataIn, accum)
+  | otherwise = (accum, accum)
+
+-- | FilterPipeline to sync all the signals
+-- contains 3 registers to determine when the filter needs to accumulate.
+-- this happens when the accumulator is ready
+filterPipeline
+  :: (KnownNat n0)
+  => (Bool, Bool, Bool)
+  -- ^ parts of the tuple
+  --
+  -- 1. accum ready with 1 cycle of delay
+  -- 2. accum ready with 2 cycles of delay
+  -- 3. latch result with 1 cycle of delay
+  -> (Bool, BitVector n0)
+  -- ^ parts of tuple
+  -- input of the mealy machine
+  -- 1. accum ready
+  -- 2. the main counter
+  -> ( (Bool, Bool, Bool),
+       (Bool, Bool, Bool))
+  -- ^ parts of the tuple
+  --
+  -- 1,2,3. input for all the registers
+  -- 4, boolean for when the filter need to accumulate
+  -- 5. boolean for when the filter accumulator of ready
+  -- 6. boolean for when the output for the adc is ready
+filterPipeline (accumRdyD1, accumRdyD2, latchResult) (accumRdy, count) =
+  ( (accumRdyD1', accumRdyD2', latchResult')
+  , (accumulate, latchResult', latchResult))
+ where
+   accumRdyD1' = accumRdy
+   accumRdyD2' = accumRdyD1
+   accumulate = accumRdyD1 && not accumRdyD2
+   latchResult' = accumulate && (count == 1)
+
+-- | Count when the input sample is ready it counts till 2^(filterWidth)
+filterCount
+  :: (HiddenClockResetEnable dom, KnownNat n0)
+  => SNat n0
+  -- ^ Filter width
+  -> Signal dom Bool
+  -- ^ Sample ready boolean
+  -> Signal dom (BitVector n0)
+  -- ^ Counter output
+filterCount _ accumRdy = output
+ where
+   output = regEn 0 accumRdy (output + 1)
+
+-- | Sigma Delta ADC configurable in the ADC width, OSR and filter depth
+--
+-- | The frequency for the ADC is equal to
+--  clk / 2^((AccumulatorWidth - BitsPerCycle - 1) + FilterDepth),
+-- where clk is the operating frequency of the ADC
+sigmaDeltaADC
+  :: ( HiddenClockResetEnable dom
+     , KnownNat n0
+     , KnownNat n1
+     , KnownNat n2
+     , KnownNat n3
+     , CLog 2 (n3 + 2) <= (n1 + n0)
+     )
+  => SNat n0
+  -- ^ ADC width
+  -> SNat (n0 + n1)
+  -- ^ Accumulator width
+  -- Has to be larger or equal to the ADC width,
+  -- the OSR is equal to 2^(AccumulatorWidth - ADCwidth)
+  -> SNat n2
+  -- ^ Filter depth
+  -- The depth of the decimation filter,
+  -- the downsampling rate is equal to 2^FilterDepth
+  -> Signal dom (BitVector (n3 + 1))
+  -- ^ analog input from the comparator
+  -- For the design either lvds or an external comparator can be used
+  -- the input signal had to be connected to the positive input while
+  -- the output from a low pass RC-network connects to the negative input.
+  -- The input is a bitvector such that ddr or a deserialiser can be utilised
+  -- to get more bits per clock cycle for an increased sampling frequency.
```

Comment:
I think this is mostly relevant documentation for the module itself or the function, but not the argument.